File: src/routes/apiroutes.ts
import { config } from "./../expoll_config"
import { ReturnCode } from "expoll-lib/interfaces"
import express, { NextFunction, Request, Response } from "express"
import adminRoutes from "./adminRoutes"
import pollRoutes from "./pollRoutes"
import userRoutes from "./userRoutes"
import voteRoutes from "./voteRoutes"
import simpleRoutes from "./simpleRoutes"
import authRoutes from "./authentication/auth"
import notificationRoutes from "./notifications/notifications"
import { compareVersion, getDataFromAny } from "../helper"

// eslint-disable-next-line new-cap
const apiRoutes = express.Router()
apiRoutes.use("/user", userRoutes)
apiRoutes.use("/poll", pollRoutes)
apiRoutes.use("/vote", voteRoutes)
apiRoutes.use("/admin", adminRoutes)
apiRoutes.use("/auth", authRoutes)
apiRoutes.use("/simple", simpleRoutes)
apiRoutes.use("/notifications", notificationRoutes)

const test = async (req: Request, res: Response, next: NextFunction) => {
    return res.status(200).json({
        message: "Test successful"
    })
}

export const metaInfo = async (req: Request, res: Response, next: NextFunction) => {
    const returnData = {
        body: req.body ?? "",
        headers: req.headers,
        cookies: req.cookies,
        signedCookies: req.signedCookies,
        url: req.url,
        path: req.path,
        method: req.method,
        protocol: req.protocol,
        route: req.route,
        params: req.params,
        hostname: req.hostname,
        ip: req.ip,
        httpVersion: req.httpVersion,
        secure: req.secure,
        subdomains: req.subdomains,
        xhr: req.xhr,
        serverInfo: {
            version: config.serverVersion,
            serverPort: config.serverPort,
            frontendPort: config.frontEndPort,
            loginLinkBase: config.loginLinkURL,
            mailSender: config.mail.mailUser
        }
    }
    return res.status(ReturnCode.OK).json(returnData)
}

export const serverInfo = async (req: Request, res: Response, next: NextFunction) => {
    const returnData = {
        version: config.serverVersion,
        minimumRequiredVersion: config.minimumRequiredClientVersion,
        serverPort: config.serverPort,
        frontendPort: config.frontEndPort,
        loginLinkBase: config.loginLinkURL,
        mailSender: config.mail.mailUser
    }
    return res.status(ReturnCode.OK).json(returnData)
}

export const checkCompliance = async (req: Request, res: Response, next: NextFunction) => {
    try {
        const clientVersion = getDataFromAny(req, "version") as string
        return res.status(ReturnCode.OK).send(`${compareVersion(clientVersion, config.minimumRequiredClientVersion)}`)
    } catch (err) {
        console.warn(err)
        return res.status(ReturnCode.INTERNAL_SERVER_ERROR).json({
            message: err
        })
    }
}

apiRoutes.get("/test", test)
apiRoutes.all("/metaInfo", metaInfo)
apiRoutes.all("/serverInfo", serverInfo)
apiRoutes.all("/compliance", checkCompliance)

export default apiRoutes

File: src/main/kotlin/net/mt32/expoll/routes/APIRoutes.kt
package net.mt32.expoll.routes

import io.ktor.http.*
import io.ktor.server.auth.*
import io.ktor.server.request.*
import io.ktor.server.response.*
import io.ktor.server.routing.*
import io.ktor.util.*
import net.mt32.expoll.auth.normalAuth
import net.mt32.expoll.commons.serializable.ServerInfo
import net.mt32.expoll.helper.defaultJSON
import net.mt32.expoll.instance
import net.mt32.expoll.iosPlatformInfo
import net.mt32.expoll.routes.admin.adminRoute
import net.mt32.expoll.routes.auth.authRoutes

fun Route.apiRouting() {
    route("/") {
        get("test") {
            call.respondText("Hello World!")
        }
        clientCompatibilityRoute()
        get("serverInfo") {
            call.respond(ServerInfo.instance)
        }
        get("ping"){
            call.respondText("pong", ContentType.Text.Plain)
        }
        route("appInfo"){
            get("ios"){
                call.respond(iosPlatformInfo)
            }
        }
        get("metaInfo") {
            call.respondText(
                """
                {
                    "body": "${call.receiveText()}",
                    "headers": ${defaultJSON.encodeToString(call.request.headers.toMap())},
                    "cookies": ${defaultJSON.encodeToString(call.request.cookies.rawCookies)},
                    "signedCookies": "req.signedCookies",
                    "url": "${call.request.uri}",
                    "path": "${call.request.uri}",
                    "method": "${call.request.httpMethod.value}",
                    "protocol": "${call.request.local.scheme}",
                    "route": "req.route",
                    "params": ${defaultJSON.encodeToString(call.parameters.toMap())},
                    "hostname": "${call.request.local.serverHost}",
                    "ip": "${call.request.local.remoteAddress}",
                    "httpVersion": "${call.request.local.version}",
                    "secure": "req.secure",
                    "subdomains": "req.subdomains",
                    "xhr": "req.xhr",
                    "serverInfo": ${defaultJSON.encodeToString(ServerInfo.instance)},
                    "localport": ${call.request.local.localPort},
                    "serverport": ${call.request.local.serverPort}
                }
            """.trimIndent(), ContentType.Application.Json
            )
        }
        authRoutes()
        simpleRoutes()
        userRoutes()
        authenticate(normalAuth) {
            pollRoutes()
            voteRoutes()
            notificationRoutes()
        }
        adminRoute()
    }
}